f5/bigip/test/test_resource.py

```diff
@@ -0,0 +1,351 @@
+        r = Resource(mock.MagicMock())
+        r._meta_data['allowed_lazy_attributes'] = []
+        r._meta_data['uri'] = 'URI'
+        r._meta_data['bigip']._meta_data['icr_session'].get.return_value =\
+            MockResponse({u"generation": 0})
+        r._meta_data['bigip']._meta_data['icr_session'].put.return_value =\
+            MockResponse({u"generation": 0})
+        r.generation = 0
+        pre_meta = r._meta_data.copy()
+        r.update(a=u"b")
+        assert pre_meta == r._meta_data
+
+    def test_Collection_removal(self):
+        r = Resource(mock.MagicMock())
+        r._meta_data['allowed_lazy_attributes'] = []
+        r._meta_data['uri'] = 'URI'
+        attrs = {'put.return_value': MockResponse({u"generation": 0}),
+                 'get.return_value': MockResponse({u"generation": 0})}
+        mock_session = mock.MagicMock(**attrs)
+        r._meta_data['bigip']._meta_data = {'icr_session': mock_session}
+        r.generation = 0
+        r.contained = Collection(mock.MagicMock())
+        assert 'contained' in r.__dict__
+        r.update(a=u"b")
+        submitted = r._meta_data['bigip'].\
+            _meta_data['icr_session'].put.call_args[1]['json']
+        assert 'contained' not in submitted
+
+    def test_read_only_removal(self):
+        r = Resource(mock.MagicMock())
+        r._meta_data['allowed_lazy_attributes'] = []
+        r._meta_data['uri'] = 'URI'
+        r._meta_data['read_only_attributes'] = [u"READONLY"]
+        attrs = {'put.return_value': MockResponse({u"generation": 0}),
+                 'get.return_value': MockResponse({u"generation": 0})}
+        mock_session = mock.MagicMock(**attrs)
+        r._meta_data['bigip']._meta_data = {'icr_session': mock_session}
+        r.generation = 0
+        r.READONLY = True
+        assert 'READONLY' in r.__dict__
+        r.update(a=u"b")
+        submitted = r._meta_data['bigip'].\
+            _meta_data['icr_session'].put.call_args[1]['json']
+        assert 'READONLY' not in submitted
+
+
+class TestResource_delete(object):
+    def test_success(self):
+        r = Resource(mock.MagicMock())
+        r._meta_data['allowed_lazy_attributes'] = []
+        r._meta_data['uri'] = 'URI'
+        attrs = {'delete.return_value':
+                 MockResponse({u"generation": 0, u"status_code": 200}),
+                 'get.return_value':
+                 MockResponse({u"generation": 0, u"status_code": 200})}
+        mock_session = mock.MagicMock(**attrs)
+        r._meta_data['bigip']._meta_data = {'icr_session': mock_session}
+        r.generation = 0
+        r.delete(force=False)
+        assert r.__dict__ == {'deleted': True}
+
+    def test_invalid_force(self):
+        r = Resource(mock.MagicMock())
+        r._meta_data['uri'] = 'URI'
+        mock_session = mock.MagicMock()
+        r._meta_data['bigip']._meta_data = {'icr_session': mock_session}
+        with pytest.raises(InvalidForceType) as IFTEIO:
+            r.delete(force='true')
```

Comment:
If "force" is ever anything besides a Python bool (i.e. `True` or `False`) then an exception is raised.  That's the purpose of this test.